File: vscode/src/ime/system/SystemImeProvider.ts
import {
  ImeAsciiModeSwitcher,
  ImeCapsLockSwitcher,
  ImeStateSource,
} from '../../core/types';
import {
  isNativeAvailable,
  nativeCapsRead,
  nativeCapsSet,
  nativeIsComposing,
  nativeSetAsciiMode,
  nativeSystemImeStatus,
} from './native';

export enum SystemType {
  WINDOWS = 'windows',
  MACOS = 'macos',
  LINUX = 'linux',
  UNKNOWN = 'unknown',
}

export function currentSystemType(platform = process.platform): SystemType {
  switch (platform) {
    case 'win32':
      return SystemType.WINDOWS;
    case 'darwin':
      return SystemType.MACOS;
    case 'linux':
      return SystemType.LINUX;
    default:
      return SystemType.UNKNOWN;
  }
}

export interface SystemImeProvider
  extends ImeStateSource, ImeAsciiModeSwitcher, ImeCapsLockSwitcher {
  readonly type: SystemType;
}

type SystemProviderFactory = () => SystemImeProvider;

export class SystemImeProviderRegistry {
  private readonly providers = new Map<SystemType, SystemProviderFactory>();

  register(type: SystemType, factory: SystemProviderFactory): void {
    this.providers.set(type, factory);
  }

  create(type: SystemType): SystemImeProvider {
    const factory = this.providers.get(type);
    if (!factory) {
      throw new Error(`System IME provider not found for type: ${type}`);
    }
    return factory();
  }

  supportedTypes(): SystemType[] {
    return [...this.providers.keys()];
  }
}

export class WindowsSystemImeProvider implements SystemImeProvider {
  readonly type = SystemType.WINDOWS;

  readAsciiMode(): boolean | null {
    return nativeSystemImeStatus()?.isAsciiMode ?? null;
  }

  readCapsLock(): boolean | null {
    return isNativeAvailable() ? nativeCapsRead() : null;
  }

  readComposing(): boolean {
    return nativeIsComposing() === 1;
  }

  async switchAsciiMode(
    ascii: boolean,
    shouldContinue: () => boolean,
  ): Promise<boolean> {
    for (let attempt = 0; attempt < 3; attempt++) {
      if (!shouldContinue()) return false;
      if (nativeSetAsciiMode(ascii)) return true;
      await delay(50);
    }
    return false;
  }

  async switchCapsLock(
    enabled: boolean,
    shouldContinue: () => boolean,
  ): Promise<boolean> {
    for (let attempt = 0; attempt < 5; attempt++) {
      if (!shouldContinue()) return false;
      if (isNativeAvailable() && nativeCapsRead() === enabled) return true;
      nativeCapsSet(enabled);
      await delay(50);
    }
    return isNativeAvailable() && nativeCapsRead() === enabled;
  }
}

function delay(ms: number): Promise<void> {
  return new Promise(resolve => setTimeout(resolve, ms));
}

File: core/src/main/kotlin/com/auto_switch_ime/core/ime/system/SystemImeProvider.kt
package com.auto_switch_ime.core.ime.system

import com.auto_switch_ime.core.ImeAsciiModeSwitcher
import com.auto_switch_ime.core.ImeCapsLockSwitcher
import com.auto_switch_ime.core.ImeStateSource

enum class SystemType {
    WINDOWS,
    MACOS,
    LINUX,
    UNKNOWN;

    companion object {
        fun current(osName: String = System.getProperty("os.name")): SystemType {
            val normalized = osName.lowercase()
            return when {
                normalized.contains("win") -> WINDOWS
                normalized.contains("mac") -> MACOS
                normalized.contains("linux") -> LINUX
                else -> UNKNOWN
            }
        }
    }
}

/** 操作系统级默认能力；不同操作系统通过独立实现扩展。 */
interface SystemImeProvider : ImeStateSource, ImeAsciiModeSwitcher, ImeCapsLockSwitcher {
    val type: SystemType
}

class SystemImeProviderRegistry {
    private val providers = mutableMapOf<SystemType, () -> SystemImeProvider>()

    fun register(type: SystemType, factory: () -> SystemImeProvider) {
        providers[type] = factory
    }

    fun create(type: SystemType): SystemImeProvider {
        val factory = providers[type]
            ?: throw IllegalStateException("System IME provider not found for type: $type")
        return factory()
    }

    fun supportedTypes(): Set<SystemType> = providers.keys.toSet()
}

class WindowsSystemImeProvider : SystemImeProvider {
    override val type: SystemType = SystemType.WINDOWS

    override fun readAsciiMode(): Boolean? = NativeImeSys.imeGetSystemStatus()?.isAsciiMode

    override fun readCapsLock(): Boolean? = NativeImeSys.imeCapsReadOrNull()

    override fun readComposing(): Boolean = NativeImeSys.imeIsComposing() == 1

    override suspend fun switchAsciiMode(
        ascii: Boolean,
        shouldContinue: () -> Boolean
    ): Boolean {
        repeat(3) {
            if (!shouldContinue()) return false
            if (NativeImeSys.imeSetAsciiMode(ascii)) return true
            Thread.sleep(50)
        }
        return false
    }

    override suspend fun switchCapsLock(
        enabled: Boolean,
        shouldContinue: () -> Boolean
    ): Boolean {
        repeat(5) {
            if (!shouldContinue()) return false
            if (NativeImeSys.imeCapsReadOrNull() == enabled) return true
            NativeImeSys.imeCapsSet(enabled)
            Thread.sleep(50)
        }
        return NativeImeSys.imeCapsReadOrNull() == enabled
    }
}
